### DISCUSSION TITLE
How to handle hybrid storage from external libraries

### DISCUSSION THREAD

**Discussion by dlpetrie** [benbjohnson/wtf#81] [Q&A] [answered]
Hello, I really appreciate your article and have been trying to implement the approach as it makes a lot of sense. I also watched the Brian Ketelsen talk that referred to it as well.

Where I am struggling right now is when a library uses a hybrid approach to storage. For example, I am using the https://github.com/alexedwards/scs library for session with redis as a backend storage. So this is using a combination of http cookies and redis for tracking the session.

I create a session interface in my domain type package
```
package app

// SessionUser represents a user who has authenticated into the application.
type SessionUser struct {
	ID        string
	Email     string
	FirstName string
	LastName  string
}

// SessionService represents a service for managing a user session.
type SessionService interface {
	Get(ctx context.Context) (*SessionUser, error)
	Save(ctx context.Context, state *SessionUser) error
	Delete(ctx context.Context) error
}
```

But where do I handle the implementation for this? Do I handle it in the `http` package because it uses cookies? The `redis` package because it uses redis? Or is it another packaged called `scs` to where scs becomes the dependency I'm referencing.

**Comment by benbjohnson** [comment, marked as answer]
> But where do I handle the implementation for this? Do I handle it in the http package because it uses cookies? The redis package because it uses redis? Or is it another packaged called scs to where scs becomes the dependency I'm referencing.

If you're only managing sessions in the `http` layer then I think you can avoid a service interface definition entirely and just use the `scs` library internally in the `http` package.

If you are managing sessions outside the `http` layer, then it doesn't matter too much about which package you place it into so long as its accessible. The main benefit to having the `SessionService` interface would be to have swappable implementations but that's not typical for most applications.
